## Project Analysis Proposal
### Pavan Nayak
### EE282



For the Final Project, I will be conducting multiple analyses on a dataset of transcriptome reads from FACS sorted tenocytes (tendon tissue producing fibroblasts) from a transgenic line of zebrafish in which the tendon marker scleraxis (scx) is labelled with the fluorescent protein mCherry, at time stages 36 hours post fertilization (hpf) and 48 hpf. These time stages represent developmental stages prior to the onset of active muscle contraction and post onset of active muscle contraction in zebrafish respectively. This experiment was conducted to better understand the global changes in tenocyte gene expression resulting from the mechanical force exerted by muscular contraction on the tendons.  

In this analysis, I will identify the number of bad reads in each fastq file (sample) as defined by reads containing 10 consecutive Ns, and compare this for each sample to the total number of reads to understand the percentage of bad reads per sample to get an idea of the quality of the data.   

Next, I will take the top ~400 Differentially Expressed genes (by p-value or False Discovery Rate) from an RNA-seq differential expression pipeline of this data and pipe these genes into a python script which converts any genes in ENSEMBL ID format into conventional gene names, and then searches each gene, along with a search term, against the NCBI Pubmed data base, and returns the number of articles returned for each search. Lastly, this output document will be piped into an R script to collect the output dataframe and utilize the ggplot2 package to plot a histogram of the number of articles found per gene/keyword search. From here, multiple search terms can be used on the same dataset and plotted to identify useful genes for different contexts within the research context of vertebrate tendon development, or any other context in which large lists of genes are returned as output. In order for the pipeline to work correctly, individual scripts need to work independently with inputs, and the bash scripts gluing the components together need to work as a unit. I believe this is possible and feasible as a final project as I have written the python script within the last few days and it currently works on its own with excel files as input and output. The next steps of the pipeline would require using the argparse python module to allow the python script to work on the command line with cmd input arguments, creating an R script to accept the output of this python script with command line arguments, and use ggplot to plot the graphs and output them to a file. Lastly, I need to build the bash scripts that take input excel files of gene lists, and pipes them through both the python and R scripts providing a smooth input/output process.  

This pipeline should ideally provide a useful tool for initial analysis of gene lists to identify useful genes not only for my particular thesis project for identification of novel genes in tendon development and mechanotransduction, but for any researcherâ€™s own research context when dealing with NGS experiments outputting large lists of genes.


